enhancements/ocp-coreos-layering/ocp-coreos-layering.md

```diff
@@ -0,0 +1,730 @@
+- Everything else largely stays the same; the MCD continues to perform config changes per node. (See: [Buildless layering support](#buildless-layering-support)). 
+
+#### Phase 1
+- Initial [preflight checks](#preflight-checks) are implemented. However, they will only be a warning and will not block image rollout since we will not (yet) have a way to bypass them.
+- Iterate on the mechanism established in Phase 0 by enabling rendered MachineConfigs to be layered into a final OS image before they are applied to a nodes’ disk by using the BuildController. (see [Create Final Pool Images](#create-final-pool-images)).
+- The BuildController will be reliant upon the OpenShift Image Builder. Alternatively, we could create our own custom solution (see: [How The Final Image Will Be Built](#how-the-final-image-will-be-built) for caveats).)
+
+#### Phase 2
+- CRD implementation is finalized. This enables the following:
+  - A cluster admin can build and stage a final image prior to rolling it out.
+  - Preflight checks can block an image rollout (assuming the documented overrides are not used).
+  - Allows us to remove our dependency on the in-cluster registry and/or ImageStreams since it gives the cluster admin a place to configure where the final image is pushed. This allows them to optionally use an in-cluster registry or ImageStream assuming they’ve enabled those features on their cluster.
+- Modify the BuildController to use customized build pods instead of OpenShift Image Builder to create the finalized OS image (assuming we did the initial implementation with OpenShift Image Builder).
+- Add support for heterogeneous clusters using BuildController.
+
+#### Phase 3
+- oc command is augmented with sub-commands to provide syntactic sugar and a more polished synchronous user experience for cluster admins around staging a new OS image, watching preflight checks pass, and rolling it out. The web UI gains an equivalent UX flow.
+
+**Note:** Rest of the enhancement proposal will focus on. For other phases, details can be found in [Future Phases](#future-phases) section.
+
+### Workflow Description
+By default, cluster admin shouldn't see any change in the behavior unless they want to override default osImageURL.
+
+In order to apply a custom OS image, cluster admin would run following steps:
+1. Creates a custom OS image derived from default RHEL CoreOS container image
+1. Pushes this image to a container registry that is reachable to the cluster
+1. Creates a MachineConfig overriding osImageURL for targeted pool
+1. MCO will update OS with content provided via MachineConfig like it does today for any other MachineConfig being applied.
+1. In case of an error, error message will be made visible for further debugging
+
+### API Extensions
+MCO will allow `osImageURL` field to be overridden via the MachineConfig.
+
+### Implementation details
+
```

Comment:
Can we add a Phase 0 header here to indicate we are covering Phase 0 impl details?

Reply:
It is not possible to add neatly just Phase0 header multiple times in the enhancement because linter throws error `MD024/no-duplicate-heading/no-duplicate-header Multiple headings with the same content [Context: "#### Phase 0"]`

So, I added a header `Phase 0 implementation`.  Also, I have explicitly callout in the note (at the end of Development Phases) that that rest of the implementation is about phase 0, so hopefully this shouldn't create any more confusion.